pkg/microservice/packager/core/service/service.go

```diff
@@ -0,0 +1,216 @@
+/*
+Copyright 2021 The KodeRover Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package service
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v3"
+
+	"github.com/koderover/zadig/pkg/microservice/packager/config"
+	"github.com/koderover/zadig/pkg/setting"
+	"github.com/koderover/zadig/pkg/tool/log"
+)
+
+// Packager ...
+type Packager struct {
+	Ctx *Context
+}
+
+type PackageResult struct {
+	ServiceName string       `json:"service_name"`
+	Result      string       `json:"result"`
+	ErrorMsg    string       `json:"error_msg"`
+	ImageData   []*ImageData `json:"image_data"`
+}
+
+func NewPackager() (*Packager, error) {
+	context, err := ioutil.ReadFile(config.JobConfigFile())
+	if err != nil {
+		return nil, err
+	}
+
+	var ctx *Context
+	if err := yaml.Unmarshal(context, &ctx); err != nil {
+		return nil, err
+	}
+
+	packager := &Packager{
+		Ctx: ctx,
+	}
+
+	return packager, nil
+}
+
+// BeforeExec ...
+func (p *Packager) BeforeExec() error {
+	log.Info("wait for docker daemon to start")
+	for i := 0; i < 120; i++ {
+		err := dockerInfo().Run()
+		if err == nil {
+			break
+		}
+		time.Sleep(time.Second * 1)
+	}
+
+	if len(p.Ctx.ProgressFile) == 0 {
+		return nil
+	}
+
+	// create log file
+	if err := os.MkdirAll(filepath.Dir(p.Ctx.ProgressFile), 0770); err != nil {
+		return errors.Wrapf(err, "failed to create progress file dir")
+	}
+	file, err := os.Create(p.Ctx.ProgressFile)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create progress file")
+	}
+	err = file.Close()
+	return errors.Wrapf(err, "failed to close progress file")
+}
+
+// Exec ...
+func (p *Packager) Exec() error {
+
+	// add docker registry auths for all registries
+	allRegistries := make([]*DockerRegistry, 0)
+	allRegistries = append(allRegistries, p.Ctx.SourceRegistries...)
+	allRegistries = append(allRegistries, p.Ctx.TargetRegistries...)
+	err := writeDockerConfig(allRegistries)
+	if err != nil {
+		return err
+	}
+
+	realTimeProgress := make([]*PackageResult, 0)
+
+	for _, image := range p.Ctx.Images {
+		cmds := p.dockerCommands(image)
+		var err error
+		for _, cmd := range cmds {
+			cmd.Stdout = os.Stdout
+			cmd.Stderr = os.Stderr
+			log.Info(strings.Join(cmd.Args, " "))
+			if err = cmd.Run(); err != nil {
```

Comment:
record the error

Reply:
break only breaks the inner loop, err is recorded in the following logic